subsys/settings/src/settings_zms.c

```diff
@@ -326,21 +516,86 @@ static int settings_zms_save(struct settings_store *cs, const char *name, const
 			}
 		}
 		settings_element.previous_hash = cf->last_hash_id;
-		rc = zms_write(&cf->cf_zms, name_hash | 1, &settings_element,
+		rc = zms_write(&cf->cf_zms, ZMS_LL_NODE_FROM_NAME_ID(name_hash), &settings_element,
 			       sizeof(struct settings_hash_linked_list));
 		if (rc < 0) {
 			return rc;
 		}
+
 		/* Now update the previous linked list element */
-		settings_element.next_hash = name_hash | 1;
+		settings_element.next_hash = ZMS_LL_NODE_FROM_NAME_ID(name_hash);
 		settings_element.previous_hash = cf->second_to_last_hash_id;
 		rc = zms_write(&cf->cf_zms, cf->last_hash_id, &settings_element,
 			       sizeof(struct settings_hash_linked_list));
 		if (rc < 0) {
 			return rc;
 		}
 		cf->second_to_last_hash_id = cf->last_hash_id;
-		cf->last_hash_id = name_hash | 1;
+		cf->last_hash_id = ZMS_LL_NODE_FROM_NAME_ID(name_hash);
+#ifdef CONFIG_SETTINGS_ZMS_LL_CACHE
+		if (cf->ll_cache_next < CONFIG_SETTINGS_ZMS_LL_CACHE_SIZE) {
+			cf->ll_cache[cf->ll_cache_next] = settings_element;
+			cf->ll_cache_next = cf->ll_cache_next + 1;
+		}
+#endif
+#ifdef CONFIG_SETTINGS_ZMS_NO_LL_DELETE
+no_ll_update:
+#endif /* CONFIG_SETTINGS_ZMS_NO_LL_DELETE */
+		/* Now let's write the name */
+		rc = zms_write(&cf->cf_zms, name_hash, name, strnlen(name, SETTINGS_FULL_NAME_LEN));
```

Comment:
You can re-use the output of `strnlen` in line 418, and, which is not part of this PR, there is code in line 442 that could re-use the len to shorten the path to comparison result and `strcmp` could be replaced to `memcmp`.